telethon/network/connection/tcpmtproxy.py

```diff
@@ -1,94 +1,45 @@
 import hashlib
 import os
 
-from .connection import Connection
-from .tcpabridged import AbridgedPacket
-from .tcpintermediate import IntermediatePacket, RandomizedIntermediatePacket
+from .connection import ObfuscatedConnection
+from .tcpabridged import AbridgedPacketCodec
+from .tcpintermediate import (
+    IntermediatePacketCodec,
+    RandomizedIntermediatePacketCodec
+)
 
 from ...crypto import AESModeCTR
 
 
-class TcpMTProxy(Connection):
+class MTProxyIO:
     """
-    Connector which allows user to connect to the Telegram via proxy servers
-    commonly known as MTProxy.
-    Implemented very ugly due to the leaky abstractions in Telethon networking
-    classes that should be refactored later (TODO).
-
-    .. warning::
-
-        The support for MTProtoProxies class is **EXPERIMENTAL** and prone to
-        be changed. You shouldn't be using this class yet.
+    It's very similar to tcpobfuscated.ObfuscatedIO, but the way
+    encryption keys, protocol tag and dc_id are encoded is different.
     """
-    packet_codec = None
+    header = None
 
-    @staticmethod
-    def address_info(proxy_info):
-        if proxy_info is None:
-            raise ValueError("No proxy info specified for MTProxy connection")
-        return proxy_info[:2]
+    def __init__(self, connection):
+        self._reader = connection._reader
+        self._writer = connection._writer
 
-    def __init__(self, ip, port, dc_id, *, loop, loggers, proxy=None):
-        proxy_host, proxy_port = self.address_info(proxy)
-        super().__init__(
-            proxy_host, proxy_port, dc_id, loop=loop, loggers=loggers)
-        self._codec = self.packet_codec()
-        secret = bytes.fromhex(proxy[2])
+        (self.header,
+         self._encrypt,
+         self._decrypt) = self.init_header(
+             connection._secret, connection._dc_id, connection.packet_codec)
+
+    def init_header(self, secret, dc_id, packet_codec):
+        # Validate
         is_dd = (len(secret) == 17) and (secret[0] == 0xDD)
-        if is_dd and (self.packet_codec != RandomizedIntermediatePacket):
+        is_rand_codec = (
+            packet_codec == RandomizedIntermediatePacketCodec)
```

Comment:
Maybe it's more appropriated to use `issubclass(packet_codec, RandomizedIntermediatePacketCodec)`?

Reply:
@Lonami yep, good suggestion!

Reply:
@Lonami fixed!